File: c++.stl.f
5 STL - 常用算法
概述
算法主要是由头文件<algorithm> <functional> <numeric>组成。
<algorithm>是所有STL头文件中最大的一个，范围涉及到比较、 交换、查找、遍历操作、复制、修改等等
<numeric>体积很小，只包括几个在序列上面进行简单数学运算的模板函数
<functional>定义了一些模板类, 用以声明函数对象。
5.1 常用遍历算法
学习目标：
掌握常用的遍历算法
算法简介：
for_each     //遍历容器
transform   //搬运容器到另一个容器中
5.1.1 for_each
功能描述：
实现遍历容器
函数原型：
for_each(iterator beg, iterator end, _func);  
 遍历算法 遍历容器元素
 beg 开始迭代器
 end 结束迭代器
 _func 函数或者函数对象
#include<iostream>
using namespace std;
#include<vector>
#include<algorithm>
//普通函数
void print01(int val){
	cout << val << " ";
}
//仿函数
class print02{
public:
	void operator()(int val){
		cout << val << " ";
	}
};
void test01(){
	vector<int>v;
	for (int i = 0; i < 10; i++){
		v.push_back(i);
	}
	for_each(v.begin(), v.end(), print01);
	cout << endl;
	for_each(v.begin(), v.end(), print02());
}
int main(){
	test01();
	return 0;
}

File: c++.stl.c
4.3 内建函数对象
4.3.1 内建函数对象意义
概念：
STL内建了一些函数对象
分类:
* 算术仿函数
* 关系仿函数
* 逻辑仿函数
用法：
* 这些仿函数所产生的对象，用法和一般函数完全相同
* 使用内建函数对象，需要引入头文件 #include<functional>
4.3.2 算术仿函数
功能描述：
* 实现四则运算
* 其中negate是一元运算，其他都是二元运算
仿函数原型：
template<class T> T plus<T>                //加法仿函数
template<class T> T minus<T>              //减法仿函数
template<class T> T multiplies<T>    //乘法仿函数
template<class T> T divides<T>         //除法仿函数
template<class T> T modulus<T>         //取模仿函数
template<class T> T negate<T>           //取反仿函数
#include<iostream>
using namespace std;
#include<functional>//内建函数对象头文件
void test01(){
	negate<int>n;
	cout << n(50) << endl;
}
void test02(){
	plus<int>p;
	cout << p(10, 20) << endl;
}
int main(){
	test01();
	test02();
	return 0;
}
总结：使用内建函数对象时，需要引入头文件 #include <functional>
